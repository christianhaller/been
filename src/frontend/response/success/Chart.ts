import type { EnhancedPinList } from "../../../backend/interace";

export class Chart {
  private places: EnhancedPinList;
  private el: HTMLElement;
  private chart: google.visualization.GeoChart;

  async load() {
    return new Promise((resolve) => {
      google.charts.load("current", {
        packages: ["geochart"],
        mapsApiKey: "AIzaSyCsIs5SJoutxcR0Pla5bq7lFVcAW_rr17Q",
      });
      google.charts.setOnLoadCallback(resolve);
    });
  }

  async init(parent: HTMLElement, places: EnhancedPinList) {
    if (!google) {
      throw "lib not loaded";
    }
    this.el = parent.querySelector("#chart");
    this.places = places;
    await this.load();

    const data = new google.visualization.DataTable();
    data.addColumn("number", "Lat");
    data.addColumn("number", "Long");
    data.addColumn("number", "Value");
    data.addColumn({ type: "string", role: "tooltip" });

    this.places.forEach(({ lat, lng, city, flags }) => {
      data.addRows([[lat, lng, flags.includes("been") ? 100 : 200, city]]);
    });

    const options = {
      backgroundColor: {
        fill: "transparent",
        stroke: "#FFF",
        strokeWidth: 0,
      },
      datalessRegionColor: "#f5f5f5",
      displayMode: "markers",
      resolution: "countries",
      sizeAxis: { minValue: 1, maxValue: 200 },
      legend: "none",
      keepAspectRatio: true,
      width: this.el.offsetWidth - 40,
      colorAxis: { minValue: 1, maxValue: 2, colors: ["#ff0099", "#000"] }, // orange to blue
    };

    this.chart = new google.visualization.GeoChart(this.el);

    // @ts-ignore
    this.chart.draw(data, options);
    return this;
  }
  async getImage(): Promise<Blob> {
    return new Promise((resolve) => {
      google.visualization.events.addListener(this.chart, "ready", async () => {
        const res = await fetch(this.chart.getImageURI());
        resolve(res.blob());
      });
    });
  }
}
