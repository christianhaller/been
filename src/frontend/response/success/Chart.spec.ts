import { Chart } from "./Chart.ts";

jest.mock("loadjs", () => {
  return jest.fn(() => {});
});

describe("Chart", () => {
  window.google = {
    charts: {
      load: jest.fn(),
      setOnLoadCallback: jest.fn(),
    },
  };

  test("init", () => {
    new Chart().init(document, {
      language: "en",
      username: "christianhaller",
      places: [
        {
          lat: 50.444885,
          lng: 30.536667,
          flags: ["been"],
          city: "Kyiv(Kiev)",
          country: "Ukraine",
        },
        {
          lat: -34.45451,
          lng: -57.83755,
          flags: ["been"],
          city: "Colonia del Sacramento",
          country: "Uruguay",
        },
      ],
    });
  });
});
