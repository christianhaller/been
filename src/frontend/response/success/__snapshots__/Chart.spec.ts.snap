// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chart draw 1`] = `
Array [
  Object {
    "addColumn": [MockFunction] {
      "calls": Array [
        Array [
          "number",
          "Lat",
        ],
        Array [
          "number",
          "Long",
        ],
        Array [
          "number",
          "Value",
        ],
        Array [
          Object {
            "role": "tooltip",
            "type": "string",
          },
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
        Object {
          "type": "return",
          "value": undefined,
        },
        Object {
          "type": "return",
          "value": undefined,
        },
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    },
    "addRows": [MockFunction] {
      "calls": Array [
        Array [
          Array [
            Array [
              50.444885,
              30.536667,
              1,
              "Kyiv(Kiev)",
            ],
          ],
        ],
        Array [
          Array [
            Array [
              -34.45451,
              -57.83755,
              1,
              "Colonia del Sacramento",
            ],
          ],
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    },
  },
  Object {
    "backgroundColor": Object {
      "fill": "transparent",
      "stroke": "#FFF",
      "strokeWidth": 0,
    },
    "colorAxis": Object {
      "colors": Array [
        "#ff0099",
        "#000",
      ],
      "maxValue": 10,
      "minValue": 1,
    },
    "datalessRegionColor": "#f5f5f5",
    "displayMode": "markers",
    "keepAspectRatio": true,
    "legend": "none",
    "resolution": "countries",
    "sizeAxis": Object {
      "maxValue": 200,
      "minValue": 1,
    },
    "width": -40,
  },
]
`;
